orm/serial_bucket.go

```diff
@@ -0,0 +1,384 @@
+package orm
+
+import (
+	"reflect"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+)
+
+// SerialModel is implemented by any entity that can be stored using SerialModelBucket.
+// GetID/SetID are used to store and access the Key.
+// The ID is always set to nil before serializing and storing the Value.
+type SerialModel interface {
+	weave.Persistent
+	Validate() error
+	GetID() []byte
+	SetID([]byte) error
+}
+
+// SerialModelSlicePtr represents a pointer to a slice of SerialModels. Think of it as
+// *[]SerialModel Because of Go type system, using []SerialModel would not work for us.
+// Instead we use a placeholder type and the validation is done during the
+// runtime.
+type SerialModelSlicePtr interface{}
+
+// SerialModelBucket is implemented by buckets that operates on SerialModels rather than
+// Objects.
+type SerialModelBucket interface {
```

Comment:
Is there a need for an interface here?
Rather we have ONE concrete implementation of this interface.

You can make the interface `serialModuleBucket` and the class `SerialModelBucket`.
This will define the contract and you can enforce that the struct implements the interface with such an compile-time assert. But it is nicer to return a struct and accept an interface, unless you really need to support multiple implementations.

Reply:
Nevermind, I just realized `migrations.SerialModelBucket`.

But that is a bit confusing with names, one a interface, one a struct

Maybe make `{orm,migrations}.SerialModelBucket` a struct in both cases.
And call the interface: `orm.SerialModelBucketer` (or `ISerialModelBucket` :scream: )